hmrc-mongo-common/src/main/scala/uk/gov/hmrc/mongo/throttle/ThrottledCollection.scala

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2019 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.mongo.throttle
+
+import com.google.inject.{Inject, Singleton}
+import org.mongodb.scala.{MongoClient, Observable, Observer, SingleObservable, Subscription}
+import play.api.{Configuration, Logger}
+
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.concurrent.duration.{Duration, DurationInt}
+import scala.util.{Failure, Success}
+
+
+/** Single instance to ensure same throttle is applied across all mongo queries */
+@Singleton
+class ThrottleConfig @Inject()(configuration: Configuration) {
+
+  val maxWaitQueueSize = {
+    val mongoUri = configuration.get[String]("mongodb.uri")
+    val client = MongoClient(uri = mongoUri)
+    client.settings.getConnectionPoolSettings.getMaxWaitQueueSize
+  }
+
+  /** size should be no larger than the WaitQueueSize (default size is 500)
+    * larger than this will cause `com.mongodb.MongoWaitQueueFullException: Too many operations are already waiting for a connection. Max number of operations (maxWaitQueueSize) of 500 has been exceeded.`
+    */
+  val throttleSize =
+    configuration.getOptional[Int]("mongodb.throttlesize").getOrElse(maxWaitQueueSize)
+
+  Logger.debug(s"Throttling mongo queries using throttleSize=$throttleSize")
+
+  val throttledEc =
+    ExecutionContext.fromExecutorService(
```

Comment:
Do we need to register a stopHook to ensure that the underlying thread pool is shut down cleanly upon termination?
In particular, if there are still jobs in the queue, would we want to try and delay shutdown for a while in an attempt to complete those tasks?  

Am conscious that these are typical when working in Java with `ExecutorService` directly.  Not sure if the `ExecutionContext` wrapper is handling any of this.